modules/pricer/src/main/java/com/opengamma/strata/pricer/impl/option/SabrExtrapolationRightFunction.java

```diff
@@ -0,0 +1,598 @@
+/**
+ * Copyright (C) 2015 - present by OpenGamma Inc. and the OpenGamma group of companies
+ *
+ * Please see distribution for license.
+ */
+package com.opengamma.strata.pricer.impl.option;
+
+import java.util.Arrays;
+import java.util.function.Function;
+
+import com.opengamma.strata.basics.PutCall;
+import com.opengamma.strata.basics.value.ValueDerivatives;
+import com.opengamma.strata.collect.ArgChecker;
+import com.opengamma.strata.collect.array.DoubleArray;
+import com.opengamma.strata.collect.array.DoubleMatrix;
+import com.opengamma.strata.math.impl.linearalgebra.DecompositionResult;
+import com.opengamma.strata.math.impl.linearalgebra.SVDecompositionCommons;
+import com.opengamma.strata.math.impl.rootfinding.BracketRoot;
+import com.opengamma.strata.math.impl.rootfinding.RidderSingleRootFinder;
+import com.opengamma.strata.pricer.impl.volatility.smile.function.SabrFormulaData;
+import com.opengamma.strata.pricer.impl.volatility.smile.function.SabrHaganVolatilityFunctionProvider;
+import com.opengamma.strata.pricer.impl.volatility.smile.function.VolatilityFunctionProvider;
+
+/**
+ * Pricing function in the SABR model with Hagan et al. volatility function and controlled extrapolation 
+ * for large strikes by extrapolation on call prices.
+ * <p>
+ * The form of the extrapolation as a function of the strike is
+ * \begin{equation*}
+ * f(K) = K^{-\mu} \exp\left( a + \frac{b}{K} + \frac{c}{K^2} \right).
+ * \end{equation*}
+ * <P>
+ * Benaim, S., Dodgson, M., and Kainth, D. (2008). An arbitrage-free method for smile extrapolation. Technical report, Royal Bank of Scotland.
+ * <P>
+ * OpenGamma implementation note: Smile extrapolation, version 1.2, May 2011.
+ */
+public class SabrExtrapolationRightFunction {
+  /**
+   * The volatility provider. 
+   */
+  private final VolatilityFunctionProvider<SabrFormulaData> sabrFunction;
+  /**
+   * The forward.
+   */
+  private final double forward;
+  /**
+   * The time to option expiry.
+   */
+  private final double timeToExpiry;
+  /**
+   * The SABR parameter for the pricing below the cut-off strike.
+   */
+  private final SabrFormulaData sabrData;
+  /**
+   * The cut-off strike. The smile is extrapolated above that level.
+   */
+  private final double cutOffStrike;
+  /**
+   * The tail thickness parameter.
+   */
+  private final double mu;
+  /**
+   * The three fitting parameters.
+   */
+  private final double[] parameter;
+  /**
+   * The derivative of the three fitting parameters with respect to the forward.
+   * <p>
+   * Those parameters are computed only when and if required.
+   */
+  private double[] parameterDerivativeForward = new double[3];
+  /**
+   * Flag indicating if the parameter derivatives to forward have been computed.
+   */
+  private boolean parameterDerivativeForwardComputed;
```

Comment:
This does not follow concurrent programming rules. Is calculating these extra fields expensive? If not, just calculate them always. If not, then the field has to be declared volatile and accessed using the correct double checked locking style (which has to be written carefully). Note that there is no need for the boolean as you can check for null.